src/Util/PhpServicesCreator.php

```diff
@@ -0,0 +1,610 @@
+                        foreach ($value as $argValue) {
+                            $this->addLineStmt($this->createTagMethod($argValue));
+                        }
+                        break;
+                    case 'calls':
+                        foreach ($value as $argValue) {
+                            $this->addLineStmt($this->createCallMethod($argValue));
+                        }
+                        break;
+                    default:
+                        $this->addLineStmt($this->createMethod($method, $this->createSequentialArray($value)));
+                        break;
+                }
+
+                continue;
+            }
+
+            if ('arguments' === $method || 'bind' === $method) {
+                foreach ($value as $argKey => $argValue) {
+                    $this->addLineStmt($this->createMethod(
+                        'arg',
+                        $this->createStringArgument($argKey),
+                        $this->toString($argValue)
+                    ));
+                }
+
+                continue;
+            }
+
+            throw new \Exception(sprintf('Unexpected service configuration option: "%s". This could be a valid config option that the converter does not support yet.', $method));
+        }
+    }
+
+    private function createMethod(string $method, ...$argumentStrings): string
+    {
+        if ('public' === $method && 'false' === $argumentStrings[0]) {
+            $method = 'private';
+            $argumentStrings[0] = null;
+        }
+
+        if ('arguments' === $method) {
+            $method = 'args';
+        }
+
+        return $this->tab().sprintf('->%s(%s)', $method, implode(', ', $argumentStrings));
+    }
+
+    private function createDecorateMethod(array $value): string
+    {
+        $innerName = isset($value['decoration_inner_name'])
+            ? ', '.$this->toString($value['decoration_inner_name'])
+            : ', null';
+
+        $priority = isset($value['decoration_priority'])
+            ? ', '.$this->toString($value['decoration_priority'])
+            : ', 0';
+
+        $invalid = isset($value['decoration_on_invalid'])
+            ? $value['decoration_on_invalid']
+            : null;
+
+        if ($invalid) {
+            $this->addUseStatementIfNecessary(ContainerInterface::class);
+            $invalid = 'exception' === $invalid
+                ? ', ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE'
+                : ', ContainerInterface::IGNORE_ON_INVALID_REFERENCE';
+        }
+        // Don't write the next arguments if they are null.
+        if (null === $invalid && ', 0' === $priority) {
+            $priority = null;
+            $innerName = ', null' === $innerName ? null : $innerName;
+        }
+
+        return $this->tab().'->decorate('.$this->createStringArgument($value['decorates']).$innerName.$priority.$invalid.')';
+    }
+
+    private function createCallMethod(array $values): string
```

Comment:
This currently handles this syntax (from your test):

```yml
services:
  Symfony\Bundle\MakerBundle\Tests\Util\yaml_php_convert_fixtures\FakeClass\SimpleFakeClassService:
    calls:
      - [withMailer, ['@mailer'], false]
      - [setLogger, ['@logger']]
      - [setMailer, ['@mailer']] 
```

But there is 1-2 more syntaxes:

```yml
services:
  Symfony\Bundle\MakerBundle\Tests\Util\yaml_php_convert_fixtures\FakeClass\SimpleFakeClassService:
    calls:
      - { method: withMailer, arguments: ['@mailer'], returns_clone: false }
      - setLogger: [@logger']
      # and there is also a "TaggedValue" allowed ONLY if you use this last syntax
      - withLogger: !returns_clone ['@logger']
```

It's, honestly, super tricky. If it helps, here is how the logic looks in core: https://github.com/symfony/symfony/blob/f9411ab4806f7f669deb4bea927cd2f251411c52/src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php#L525-L550

Reply:
```php
private function createTaggedValue($value): string
    {
        if ('service' === $value->getTag()) {
            $className = $this->addUseStatementIfNecessary($value->getValue()['class']);

            return 'inline('.$className.')';
        }

        if (\is_array($value->getValue())) {

            $arrayForList = [];

            if (isset($value->getValue()['tag'])) {
                $arrayForList['tag'] = $value->getValue()['tag'];
            }

            if (isset($value->getValue()['index_by'])) {
                $arrayForList['index_by'] = $value->getValue()['index_by'];
            }

            if (isset($value->getValue()['default_index_method'])) {
                $arrayForList['default_index_method'] = $value->getValue()['default_index_method'];
            }

            $args = $this->createListFromArray($arrayForList);
        } else {
            $args = $this->toString($value->getValue());
        }

        return $value->getTag().'('.$args.')';
    }

Reply:
Like you like now.